### ISSUE TITLE
many examples use deprecated imports

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5074]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)| @jerub reported|
|-|-|
|Trac ID|trac#1637|
|Type|defect|
|Created|2006-04-12 22:04:54Z|

twisted.protocols.smtp and twisted.protocols.imap4 is now twisted.mail.smtp and twisted.mail.imap4.

See attached patch.

Attachments:

* [maildocpatch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3d4/3d4bce24d3e7869a1459fd4cd1bfe131f5ad2a81/d432d4ca1797f9b45f07cff44e678f669ebb7626.diff) (826 bytes) - added by jerub on 2006-04-12 22:06:00Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1637 1637
type__defect defect
reporter__jerub jerub
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__None None
time__1144879494000000 1144879494000000
changetime__1474197555542286 1474197555542286
version__ 
owner__jerub jerub

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub<br>@jerub set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jerub|
|-|-|

Patch looks good, so far as it goes.  Perhaps there are other occurrences of this in the repository?  It'd be nice to fix them all.  Can you check and update the patch if you find any?

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

There's one or two. I'm going to patch what I can see...
```
$ find . -name \*.py | xargs grep twisted.protocols  | wc -l
      53
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

How to hunt down these bugs:
```
import os
for line in os.popen('find . -name \*.py | xargs grep twisted.protocols', 'r'):
    fname, code = line.strip().split(':', 1)
    code = code.strip()
    try:
        exec code
    except DeprecationWarning:
        print fname, '->', code
    except:
        pass
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

branch is maildoc-1637 - there's a heap of changes, and the script snippet above (which must be run with python -Werror) will come up clean except for twisted.protocols.telnet, which ./twisted/manhole/telnet.py and  ./twisted/test/test_import.py both still use.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight set owner to @jerub|
|-|-|

Looks good, ship it.

**Comment by twisted-trac** [comment]
|<img alt="stephen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephen set status to `closed`|
|-|-|

(In [17455]) Remove deprecated 'from twisted.protocols import foo' imports from our examples.

Authors: jerub
Reviewer: exarkun
Closes [#5074](https://github.com/twisted/twisted/issues/5074) 

A quick little script was used to find imports from 'twisted.protocols', then
go through and import those files and see if a DeprecationWarning comes up.

import os
for line in os.popen('find . -name \*.py | xargs grep twisted.protocols', 'r'):
    fname, code = line.strip().split(':', 1)
    code = code.strip()
    try:
        exec code
    except DeprecationWarning:
        print fname, '->', code
    except:
        pass

By this method, a large number of deprecated imports were discovered and
corrected.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
